Automatic peripheral clock management

ppannuto opened the pull request:
### Pull Request Overview

This introduces a new mechanism for handling MMIO.

Nominally, every peripheral has an array of MMIO registers that live at some address in memory. The creation of a Rust object to access these registers is an unsafe operation, as it binds an arbitrary pointer to the a registers structure. Critically, this transformation was the _only_ use of unsafe in the chips crate, so by moving the creation step to the kernel, we should be able to phase out the use of unsafe in all of chips.

The next idea is the `MMIOClockInterface`. The idea here is that most peripherals have a clock that:

  - Must be enabled before accessing the peripheral
  - Must be left enabled while the peripheral is doing something (or waiting for something!)
  - Should be disabled whenever the peripheral is not in use

Since we've encapsulated access to the MMIO registers for the peripheral, we can enforce these by tying the ownership semantics of the MMIO registers to the clocks. Specifically, so long as the kernel knows how to enable and disable clocks (aka the MMIOClockInterface), before handing back a reference to the peripheral, the kernel will ensure the clock is on. Whenever peripheral access is complete, the drop method will then turn the clock off if it can be turned off. Note that the check to possibly disable the clock is unique to each peripheral.

### Testing Strategy

Only tested on hails. Predominantly by testing various USART-only and/or I2C test apps. I've verified with measurements that I2C goes into a low-power state between operations, while ping-ponging between I2C master and slave. (aside: This interface was critical to tracking down bugs to make this work, there was an ugly path in the interrupt handlers that tried to enable both TWIM and TWIS clock that Rust caught at _compile-time_, very cool)

### TODO or Help Wanted

The USART has a custom manager to deal with supporting `panic!`. I have some WIP stuff to clean up panics that should hopefully make this cleaner as well.

There are some other issues that impeded low-power I2C operation that I've pulled out to a separate PR (coming momentarily) as they're really orthogonal.

### Documentation Updated

- [x] Kernel: The relevant files in `/docs` have been updated or no updates are required.
- [x] ~~Userland: The application README has been added, updated, or no updates are required.~~

### Formatting

- [x] `make formatall` has been run.

-----

[Rendered](https://deploy-preview-760--docs-tockosorg.netlify.com/kernel/common/peripherals/index.html)

ppannuto commented:
Josh pointed out that because USART has some funkiness with panic and I2C has the complications of intermingled clocks, it's a bit harder to see what this looks like for a "normal" peripheral, so I just adapted the SPI.

I've validated via a (lightly modified) form of the spi_byte test app that hail correctly goes into low power state while SPI is active. As a general note, it took ~50m end-to-end to adapt SPI, largely dominated by adapting the interface to allow the clock guards access to the hardware as well as the registers, which is probably how it should have been architected all along.

phil-levis commented:
We were going to spend some time next week talking about power management more generally -- could you be ready to present and talk about this for 15m or so?

ppannuto commented:
Can do

niklasad1 commented:
Cool with procedural macros, but this will take a while to understand though not really familiar with them.

ppannuto commented:
I tried to write pretty comprehensive documentation for this, should be a good place to start:
https://deploy-preview-760--docs-tockosorg.netlify.com/kernel/common/mmio/index.html

@phil-levis could we add reviewing that documentation to the pre-call agenda for next week?

alevy commented:
@ppannuto I'm really excited about this, I'll try to sit down with the code and documentation in the next couple days. I also hope you're aware that you've implicitly signed up for writing a blog post about this!

bradjc commented:
This is neat. My two thoughts:

1. We need a "marketing" name. `derive-no-clock-control` doesn't sound very exciting.

2. This would make dependencies real in Tock. Now they are in the kernel, and unsafe is allowed (and is used in at least two of the dependencies), but it certainly makes the audit space larger.

bradjc commented:
Takeaways from the discussion:

1. We might want to bring the `unsafe` back in to the chips crate so that auditors can easily find it. Amit is working on this.
2. We agreed to remove the macro for generating "no-op" code for now (and the related crate and dependencies) and wait until the needed features are merged into rust core.

ppannuto commented:
re 2: I left the implementation of the derive in the PR, I just removed its use (it's a separate crate and I removed that crate from the sam4l cargo.toml). This way we're not pulling in external deps, but the code is there / ready when we want it / can use it

ppannuto commented:
@amit I like this! I was concerned from the discussion on the call that all of the invocations of MMIOManger::new were going to need unsafe, which would have been much less ergonomic, but this is great. I amended your commit slightly to fix a `cost` -> `const` typo in the documentation :D

I also changed the SpiHw to save 4 bytes (!) by dropping the mmio_address from the structure where it  wasn't needed (really I did this so there was an example implemented that matches the documentation suggestion that the Hw struct doesn't need to hold the mmio_address)

alevy commented:
@ppannuto cool. My handle is @alevy, though ;)

ppannuto commented:
Can we start the 1-week clock on this? I think it's ready to go.

(also, apologies to the other amit for the accidental tag)

bradjc commented:
My two cents: I'd shy away from the word "automatic" just because that gave me the impression that the kernel was somehow taking care of everything, when in fact the peripheral driver author is the one that has to implement exactly what happens to manage the clocks or whatever else (including knowing whether to keep the clock on or not). `PeripheralManager` or `PowerGuard` or `HardwareManager` seem like better fits to me.

ppannuto commented:
I think I've addressed all of these. I went with `PeripheralManagement` as the trait name.

The usart conflict is easy to fix, came up after #766 merged (it was always going to be a race..), I'll deal with it in the morning.

ppannuto commented:
I think this should be good to go again

bradjc commented on kernel/derive-no-peripheral-management/README.md:
```diff
@@ -0,0 +1,59 @@
+
+ - Update your Cargo.toml:
+
+    derive-no-clock-control = { path = "../../kernel/derive-no-clock-control" }
```
derive-no-peripheral-management

ppannuto commented:
> Why would a peripheral use the no-op crate rather than just not use this at all 

It would still minimize the use of unsafe. Instead of creating regs objects in every function, you would only have the one unsafe StaticRef creation

ppannuto commented:
True. The other motivation that comes to mind is that it’s easy to add
hooks in the future if you decide you do want to manage something about the
peripheral. Also makes it a nice stepping stone bit for development, start
by getting things working / always on, then start optimizing power without
a giant refactor.
On Fri, Mar 16, 2018 at 10:40 Amit Levy <notifications@github.com> wrote:

> you would only have the one unsafe StaticRef creation
>
> But you don't need to use the MMIO manager to use StaticRef
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/helena-project/tock/pull/760#issuecomment-373790569>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUt3lAZWdkVD4mW51MXU5kV4nGwPYr8ks5te_kagaJpZM4SWJoG>
> .
>

brghena commented on chips/sam4l/src/usart.rs:
```diff
@@ -276,6 +276,59 @@ const USART_BASE_ADDRS: [*mut UsartRegisters; 4] = [
+
+        let rx_active = self.rx_dma.map_or(false, |rx_dma| rx_dma.is_enabled());
+        let tx_active = self.tx_dma.map_or(false, |tx_dma| tx_dma.is_enabled());
+        if !(rx_active || tx_active || ints_active || self.is_panic) {
```
Why do we not want to disable the clock just because we are in a `panic!`? I agree that it is pointless to disable the clock, but it seems harmless to do so if the other checks pass. If there is an important reason, please add some comments to document it.

brghena commented on kernel/derive-no-peripheral-management/README.md:
```diff
@@ -0,0 +1,63 @@
+}
+
+/// The Tock object that holds all information for this peripheral.
+#[derive(NoPeripheralManagement)]
```
Along the same thread as Brad and Amit's comments, I'm unsure what the purpose of deriving NoPeripheralManagement is. It's super neat that it's possible. However, it feels like in practice things should add peripheral management the normal way and just leave some of the functions as no-ops. What's the advantage of this over that?

ppannuto replied:
This is _exactly_ that, only you don't have to write the empty no-op functions yourself, the derive writes them for you.

ppannuto commented on boards/imix/src/i2c_dummy.rs:
```diff
@@ -86,7 +86,7 @@ impl hil::i2c::I2CHwMasterClient for AccelClient {
                 buffer[0] = 0x2A as u8; // CTRL_REG1
                 buffer[1] = 1; // Bit 1 sets `active`
-                dev.write(0x1e, i2c::START | i2c::STOP, buffer, 2);
+                dev.write(0x1e, buffer, 2);
```
Unclear honestly. I don't know how (if?) that was compiling before, but no, the I2C interface did not change.

ppannuto commented on kernel/src/platform/mod.rs:
```diff
@@ -23,3 +23,26 @@ pub trait Chip {
+}
+
+/// Helper struct for interfaces that expect clocks, but have no clock control
+pub struct NoClockControl {}
```
The deeper problem is actually I2C, where there's one `I2CHw` object but two clocks of interest. So you end up with:
`impl PeripheralManagement<TWIMClock> for I2CHw { ...`
`impl PeripheralManagement<TWISClock> for I2CHw { ... `
in order for the right thing to happen based on whether you're doing TWIM or TWIS things. This means that the PeripheralManagement has to be abstract across a `<ClockInterface>`. But what about things that don't have clocks? They can use `NoClockControl` to satisfy the `<ClockInterface>` constraint

alevy commented:
I think "Peripheral" management is misleading, because we often use _peripheral_ to refer to other chips on the same board, rather than HW controllers on the MCU.

brghena commented:
I agree that it is annoying that peripheral is used generically for a lot of things. However, peripheral is used heavily in the embedded community to mean controllers within a microcontroller. Moreover, in Tock we have this same usage, referring to DMA peripherals, TWIM peripherals, USART peripherals, etc. (and unfortunately BLE peripherals, but I don't think someone is going to confuse that code within the `chips` folder)

alevy commented:
@brghena ok, sold

alevy commented:
@ppannuto I think the only blocking issues are @bradjc and @brghena being unsure about including the `NoPeripheralManagement` derive.

I think we really really want the rest of the changes merged ASAP, so how do you feel about removing that code from this PR and doing a separate one where we can argue till the cows come home?

I'm personally partial to having something like this derive, though it'd be good to have an example where it's useful somehow that just using `StaticRef` isn't enough.

ppannuto commented:
@alevy done, pulled the derive out to #821.
